sdk/log/exporter.go

```diff
@@ -127,3 +131,132 @@ func (e exportData) respond(err error) {
 		}
 	}
 }
+
+// bufferExporter provides asynchronous and synchronous export functionality by
+// buffering export requests.
+type bufferExporter struct {
+	Exporter
+
+	input   chan exportData
+	inputMu sync.Mutex
+
+	done    chan struct{}
+	stopped atomic.Bool
+}
+
+// newBufferExporter returns a new bufferExporter that wraps exporter. The
+// returned bufferExporter will buffer at most size number of export requests.
+// If size is less than zero, zero will be used (i.e. only synchronous
+// exporting will be supported).
+func newBufferExporter(exporter Exporter, size int) *bufferExporter {
+	if size < 0 {
+		size = 0
+	}
+	input := make(chan exportData, size)
+	return &bufferExporter{
+		Exporter: exporter,
+
+		input: input,
+		done:  exportSync(input, exporter),
+	}
+}
+
+var errStopped = errors.New("exporter stopped")
+
+func (e *bufferExporter) enqueue(ctx context.Context, records []Record, rCh chan<- error) error {
+	data := exportData{ctx, records, rCh}
+
+	e.inputMu.Lock()
+	defer e.inputMu.Unlock()
+
+	// Check stopped before enqueueing now that e.inputMu is held. This
+	// prevents sends on a closed chan when Shutdown is called concurrently.
+	if e.stopped.Load() {
+		return errStopped
+	}
```

Comment:
We can also add this (second time) at the beginning of the function to reduce the overhead of acquiring a lock when it is not needed (double-checked locking).

Reply:
I was on the fence about adding that.

Doing so will impact all calls to enqueue, the vast majority of which will not be happening after shutdown. Given those calls are going to be in the hot path, and the most important performance pathway is the OnEmit one, I tried to minimize computation time and not do the extra atomic load.

On the rarer occasions where this called after a shutdown, the performance hit of acquiring a lock prior to an atomic load does not seem like something that will impact users in a meaningful way.

Have I missed something?